docs/rfcs/018-storage-messaging-2.md

```diff
@@ -0,0 +1,163 @@
+each restart or there is a risk of corruption errors.
+
+The reason is etcd making much more than what we need; it is a fault tolerant
+store with strong consistency, but I claim all we need here is just simplest pub
+sub with best effort delivery, because
+* We already have centralized source of truth for long running data, like which
+  tlis are on which nodes  -- the console.
+* Momentary data (safekeeper/pageserver progress) doesn't make sense to persist.
+  Instead of putting each change to broker, expecting it to reliably deliver it
+  is better to just have constant flow of data for active timelines: 1) they
+  serve as natural heartbeats -- if node can't send, we shouldn't pull WAL from
+  it 2) it is simpler -- no need to track delivery to/from the broker.
+  Moreover, latency here is important: the faster we obtain fresh data, the
+  faster we can switch to proper safekeeper after failure.
+* As for WAL offloading leader election, it is trivial to achieve through these
+  heartbeats -- just take suitable node through deterministic rule (min node
+  id).  Once network is stable, this is a converging process (well, except
+  complicated failure topology, but even then making it converge is not
+  hard). Such elections bear some risk of several offloaders running
+  concurrently for a short period of time, but that's harmless.
+
+  Generally, if one needs strong consistency, electing leader per se is not
+  enough; it must be accompanied with number (logical clock ts), checked at
+  every action to track causality. s3 doesn't provide CAS, so it can't
+  differentiate old/new leader, this must be solved differently.
+
+  We could use etcd CAS (its most powerful/useful primitive actually) to issue
+  these leader numbers (and e.g. prefix files in s3), but currently I don't see
+  need for that.
+
+
+Obviously best effort pub sub is much more simpler and performant; the one proposed is
+
+## gRPC broker
+
+I took tonic and [prototyped](https://github.com/neondatabase/neon/blob/asher/neon-broker/broker/src/broker.rs) the replacement of functionality we currently use
+with grpc streams and tokio mpsc channels. The implementation description is at the file header.
+
+It is just 500 lines of code and core functionality is complete. 1-1 pub sub
+gives about 120k received messages per second; having multiple subscribers in
+different connecitons quickly scales to 1 million received messages per second.
+I had concerns about many concurrent streams in singe connection, but 2^20
+subscribers still work (though eat memory, with 10 publishers 20GB are consumed;
+in this implementation each publisher holds full copy of all subscribers). There
+is `bench.rs` nearby which I used for testing.
+
+`SkTimelineInfo` is wired here, but another message can be added (e.g. if
+pageservers want to communicate with each other) with templating.
+
+### Fault tolerance
+
+Since such broker is stateless, we can run it under k8s. Or add proxying to
```

Comment:
How much time it takes to replace failing broker? If we run single one and it fails we wont be able to start new projects, right? With etcd it is bound to election timeout, not to start of a new node. Etcd clients accept list of brokers hostnames so they can try them if something fails. How will it looks with our broker set up? 

May be a good idea to list failure modes and describe actions various components need to make in each of them

Reply:
k8s restarts failing pods instantly, so it should take several seconds to recover after a failure (if k8s is healthy and working without issues). All clients (safekeepers, pageservers) will retry connection until it succeeds. Broker address stays the same and points to Ingress LB – this way we don't need to have more than one broker hostname, ingress+k8s will handle requests to the right container internally.

I would expect broker to fail only if there are issues with our k8s cluster, because broker itself is simple and doesn't store anything on disk or in memory. Regarding k8s issues, SRE team should know more about typical failures and recovery time.

> May be a good idea to list failure modes and describe actions various components need to make in each of them

+1

Reply:
> k8s restarts failing pods instantly,

It is quite tricky, how it is configured? If it is alive, how many missed healthcheck requests it will take? In theory it can even answer to healthchecks but have something jammed inside so it cannot process requests.

In multi node setup in etcd if node is stuck I would expect that after timeout client can connect to other nodes. Do we need to make it work with our solution?

Without well defined failure model it is easy to imagine weird cases, but hard to answer whether we need to cover each particular one or it is enough to cover only ones that are for some reason in our situation are more important